### ISSUE TITLE
Multiply-defined symbols linker error

### ISSUE THREAD

**Issue by steronydh** [mapbox/gzip-hpp#21]
I've included gzip.hpp in two CPP files in a VS2017 project, and when it links I get "already defined" errors for gzip::compress and gzip::decompress. Is there a way to allow multiple inclusion or do I need to include it just once in a wrapper class?

**Comment by steronydh** [comment]
I've been able to get around this by:

1 - adding `#pragma once` to all of the gzip-hpp headers

2 - replacing one `#include <gzip.hpp>` with:
```
#include <zlib.h>
#include <gzip/utils.hpp>
namespace gzip {
    // Compress method that takes a pointer an immutable character sequence (aka a string in C)
    extern std::string compress(const char* data,
        std::size_t size,
        int level = Z_DEFAULT_COMPRESSION,
        int strategy = Z_DEFAULT_STRATEGY);
}
```

**Comment by springmeyer** [comment]
Thanks, yes, this needs fixed - we'll plan to use `#pragma once` in all places necessary (and `inline` to mark the functions to avoid a violation of the ODR (https://github.com/mapbox/cpp/blob/master/glossary.md#one-definition-rule).

**Comment by springmeyer** [comment]
@GretaCB assigned you to take a look at this. This is the same problem as @kkaefer fixed in https://github.com/mapbox/vector-tile/pull/19. We should be able to trigger it (replicate it in the gzip-hpp unit tests) by breaking the tests into two .cpp files. Then when those two `.cpp` files are linked the linker will hit this error. That would simulate the problem where another program includes `gzip-hpp` in at least two places in their program and then hits this problem. As far as how to break apart the tests - it does not matter as far as replicating, but we should do it in a somewhat logical way. Ideas?

**Comment by springmeyer** [comment]
nice description of how ODR interplays with templates and definitions: https://github.com/nodejs/nan/pull/234#issuecomment-70082225. TODO: make sure this is reflected correctly in the glossary: https://github.com/mapbox/cpp/blob/master/glossary.md#one-definition-rule

**Comment by springmeyer** [comment]
ff65c47 replicates this (when `inline` is missing from the header). But this is fixed in, so will be resolved once #22 merges.

**Comment by springmeyer** [comment]
fixed by #22
